virtcontainers/api.go

```diff
@@ -931,3 +931,52 @@ func ListRoutes(ctx context.Context, sandboxID string) ([]*vcTypes.Route, error)
 
 	return s.ListRoutes()
 }
+
+// CleanupContaienr is used by shimv2 to stop and delete a container exclusively, once there is no container
+// in the sandbox left, do stop the sandbox and delete it. Those serial operations will be done exclusively by
+// locking the sandbox.
+func CleanupContainer(ctx context.Context, sandboxID, containerID string) error {
+	span, ctx := trace(ctx, "ListRoutes")
+	defer span.Finish()
+
+	if sandboxID == "" {
+		return vcTypes.ErrNeedSandboxID
+	}
+
+	lockFile, err := rwLockSandbox(ctx, sandboxID)
+	if err != nil {
+		return err
+	}
+	defer unlockSandbox(ctx, sandboxID, lockFile)
+
+	s, err := fetchSandbox(ctx, sandboxID)
+	if err != nil {
+		return err
+	}
+
+	s.stateful = false
+	s.releaseStatelessSandbox()
+
+	_, err = s.StopContainer(containerID)
+	if err != nil {
+		return err
+	}
+
+	if _, err := s.DeleteContainer(containerID); err != nil {
```

Comment:
This will fail if `containerID == ""` so I wonder if `CleanupContainer()` itself should check for that scenario as soon as possible to avoid getting to this point and then failing?